File: Algorithms/022.GenerateParentheses.py
class Solution:
    # @param {integer} n
    # @return {string[]}
    def generateParenthesis(self, n):
        result = []
        def _helper(left, right, s, n):
            if left == n and right == n:
                result.append(s)
                return
            if left < n:
                _helper(left + 1, right, s + "(", n)
            if right < left:
                _helper(left, right + 1, s + ")", n)
        _helper(0, 0, "", n)
        return result

File: Algorithms/cpp/022.GenerateParentheses.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> result;
        string s;
        generateParenthesis(n, 0, 0, s, result);
        return result;
    }
    void generateParenthesis(int n, int left, int right, string s,
            vector<string>& result) {
        if (left == n && right == n) {
            result.push_back(s);
            return;
        }
        if (left < n) {
            generateParenthesis(n, left+1, right, s + "(", result);
        }
        if (right < left) {
            generateParenthesis(n, left, right+1, s + ")", result);
        }
    }
};
